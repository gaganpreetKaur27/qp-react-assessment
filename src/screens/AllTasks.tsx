import React, { useState, useEffect, useCallback } from "react";
//components
import { TaskItem } from "../components/TaskItem";
import { CreateTask } from "../components/CreateTask";
//utils
import { TaskItemProp } from "../utils/types/TaskTypes";
import "../utils/css/styles.css";

export const AllTasks = () => {
  const [tasksRemaining, setTasksRemaining] = useState(0);
  const [tasks, setTasks] = useState([
    {
      title: "This is sample completed task",
      completed: true,
    },
    {
      title: "This is a pending task",
      completed: false,
    },
  ]);

  useEffect(() => {
    setTasksRemaining(tasks.filter((task) => !task.completed)?.length);
  }, [tasks]);

  const addTask = useCallback(
    (title: string) => {
      const newTasks = [...tasks, { title, completed: false }];
      setTasks(newTasks);
    },
    [tasks]
  );

  const completeTask = useCallback(
    (index: number) => {
      const newTasks = [...tasks];

      newTasks[index].completed = !newTasks[index].completed;
      setTasks(newTasks);
    },
    [tasks]
  );

  const removeTask = useCallback(
    (index: number) => {
      const newTasks = [...tasks];
      newTasks.splice(index, 1);
      setTasks(newTasks);
    },
    [tasks]
  );

  return (
    <div className="todo-container">
       <div className="create-task">
        <CreateTask addTask={addTask} />
      </div>
      {tasksRemaining > 0? <div className="header">Pending Items: ({tasksRemaining})</div> : 
      <div className="header">My Tasks</div>
      }
      
      <div className="tasks">
        {tasks.map((task: TaskItemProp, index: number) => (
          <TaskItem
            task={task}
            index={index}
            key={index}
            completeTask={completeTask}
            removeTask={removeTask}
          />
        ))}
      </div>
     
    </div>
  );
};
