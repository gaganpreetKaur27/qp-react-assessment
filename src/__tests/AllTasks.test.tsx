import React from "react";
import { render } from "@testing-library/react";
import { AllTasks } from "../screens/AllTasks";

describe("AllTasks component", () => {
  it("AllTasks matches snapshot", () => {
    const { asFragment } = render(<AllTasks />);
    expect(asFragment()).toMatchSnapshot();
  });
});
