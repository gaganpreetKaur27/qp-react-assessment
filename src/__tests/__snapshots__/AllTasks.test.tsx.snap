// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AllTasks component AllTasks matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="todo-container"
  >
    <div
      class="create-task"
    >
      <div
        class="add-task-container"
      >
        <input
          class="input"
          placeholder="Add a new task to the list.. :)"
          type="text"
          value=""
        />
        <button
          class="add-button-style"
        >
          Add Task
        </button>
      </div>
    </div>
    <div
      class="header"
    >
      Pending Items: (1)
    </div>
    <div
      class="tasks"
    >
      <div
        class="task"
        style="text-decoration: line-through;"
      >
        <span>
          This is sample completed task
        </span>
        <button
          class="button-container"
          style="background: red;"
        >
          x
        </button>
        <button
          class="button-container"
        >
          Completed
        </button>
      </div>
      <div
        class="task"
      >
        <span>
          This is a pending task
        </span>
        <button
          class="button-container"
          style="background: red;"
        >
          x
        </button>
        <button
          class="checkbox-button-pending"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
